File: services/network/src/tcp.rs
//! TCP protocol implementation

use core::mem;

/// TCP header structure
#[repr(C, packed)]
pub struct TcpHeader {
    pub src_port: u16,
    pub dest_port: u16,
    pub seq_number: u32,
    pub ack_number: u32,
    pub data_offset: u8,      // Data offset (4 bits) + reserved (4 bits)
    pub flags: u8,
    pub window_size: u16,
    pub checksum: u16,
    pub urgent_ptr: u16,
    pub options: [u8; 0],      // Variable length options
}

/// TCP flags
pub const TCP_FLAG_FIN: u8 = 0x01;
pub const TCP_FLAG_SYN: u8 = 0x02;
pub const TCP_FLAG_RST: u8 = 0x04;
pub const TCP_FLAG_PSH: u8 = 0x08;
pub const TCP_FLAG_ACK: u8 = 0x10;
pub const TCP_FLAG_URG: u8 = 0x20;

/// TCP states
#[repr(u32)]
pub enum TcpState {
    Closed = 0,
    Listen = 1,
    SynSent = 2,
    SynReceived = 3,
    Established = 4,
    FinWait1 = 5,
    FinWait2 = 6,
    CloseWait = 7,
    Closing = 8,
    LastAck = 9,
    TimeWait = 10,
}

/// TCP connection
pub struct TcpConnection {
    pub local_ip: u32,
    pub remote_ip: u32,
    pub local_port: u16,
    pub remote_port: u16,
    pub state: TcpState,
    pub seq_num: u32,
    pub ack_num: u32,
    pub window_size: u32,
}

const MAX_TCP_CONNECTIONS: usize = 32;
static mut TCP_CONNECTIONS: [Option<TcpConnection>; MAX_TCP_CONNECTIONS] = [None; MAX_TCP_CONNECTIONS];
static mut NEXT_SEQ_NUM: u32 = 1;
static mut INITIALIZED: bool = false;

/// Initialize TCP
pub fn tcp_init() -> Result<(), ()> {
    unsafe {
        if INITIALIZED {
            return Ok(());
        }
        
        NEXT_SEQ_NUM = 1;
        INITIALIZED = true;
        
        Ok(())
    }
}

/// Create TCP connection
pub fn tcp_create_connection(local_ip: u32, local_port: u16, remote_ip: u32, remote_port: u16) -> Result<usize, ()> {
    unsafe {
        if !INITIALIZED {
            tcp_init()?;
        }
        
        // Find free slot
        for i in 0..MAX_TCP_CONNECTIONS {
            if TCP_CONNECTIONS[i].is_none() {
                TCP_CONNECTIONS[i] = Some(TcpConnection {
                    local_ip,
                    remote_ip,
                    local_port,
                    remote_port,
                    state: TcpState::Closed,
                    seq_num: NEXT_SEQ_NUM,
                    ack_num: 0,
                    window_size: 65535,
                });
                NEXT_SEQ_NUM += 1;
                return Ok(i);
            }
        }

        Err(())
    }
}

/// Initiate TCP connection (SYN)
pub fn tcp_connect(conn_id: usize) -> Result<(), ()> {
    unsafe {
        if conn_id >= MAX_TCP_CONNECTIONS {
            return Err(());
        }

        if let Some(ref mut conn) = TCP_CONNECTIONS[conn_id] {
            conn.state = TcpState::SynSent;
            // Send SYN packet
            let mut tcp_header = TcpHeader {
                src_port: conn.local_port,
                dest_port: conn.remote_port,
                seq_number: conn.seq_num,
                ack_number: 0,
                data_offset: 0x50, // 5 * 4 = 20 bytes header
                flags: TCP_FLAG_SYN,
                window_size: conn.window_size as u16,
                checksum: 0,
                urgent_ptr: 0,
            };
            // Calculate checksum and send via IP layer
            use crate::ip::ip_send;
            let header_bytes = unsafe {
                core::slice::from_raw_parts(&tcp_header as *const _ as *const u8, 20)
            };
            let _ = ip_send(conn.remote_ip, crate::ip::IP_PROTOCOL_TCP, header_bytes);
            Ok(())
        } else {
            Err(())
        }
    }
}

/// Send data on TCP connection
pub fn tcp_send(conn_id: usize, data: &[u8]) -> Result<(), ()> {
    unsafe {
        if conn_id >= MAX_TCP_CONNECTIONS {
            return Err(());
        }

        if let Some(ref conn) = TCP_CONNECTIONS[conn_id] {
            if conn.state != TcpState::Established {
                return Err(());
            }

            // Build and send TCP segment
            let mut tcp_header = TcpHeader {
                src_port: conn.local_port,
                dest_port: conn.remote_port,
                seq_number: conn.seq_num,
                ack_number: conn.ack_num,
                data_offset: 0x50,
                flags: TCP_FLAG_ACK | TCP_FLAG_PSH,
                window_size: conn.window_size as u16,
                checksum: 0,
                urgent_ptr: 0,
            };
            // Build packet: header + data
            let mut packet = [0u8; 1500];
            unsafe {
                core::ptr::copy_nonoverlapping(&tcp_header as *const _ as *const u8, packet.as_mut_ptr(), 20);
            }
            let data_len = data.len().min(1480);
            packet[20..20+data_len].copy_from_slice(&data[0..data_len]);
            // Send via IP layer
            use crate::ip::ip_send;
            let _ = ip_send(conn.remote_ip, crate::ip::IP_PROTOCOL_TCP, &packet[0..20+data_len]);
            Ok(())
        } else {
            Err(())
        }
    }
}

/// Receive data from TCP connection
pub fn tcp_receive(conn_id: usize, buffer: &mut [u8]) -> Result<usize, ()> {
    unsafe {
        if conn_id >= MAX_TCP_CONNECTIONS {
            return Err(());
        }

        if let Some(ref conn) = TCP_CONNECTIONS[conn_id] {
            if conn.state != TcpState::Established {
                return Err(());
            }

            // Retrieve data from receive buffer
            // For now, return 0 (receive buffer not fully implemented)
            // Full implementation would:
            // 1. Check receive buffer for data
            // 2. Copy data to buffer
            // 3. Update ack_num
            let _ = buffer;
            Ok(0)
        } else {
            Err(())
        }
    }
}

/// Close TCP connection
pub fn tcp_close(conn_id: usize) -> Result<(), ()> {
    unsafe {
        if conn_id >= MAX_TCP_CONNECTIONS {
            return Err(());
        }

        if let Some(ref mut conn) = TCP_CONNECTIONS[conn_id] {
            conn.state = TcpState::FinWait1;
            // Send FIN packet
            let mut tcp_header = TcpHeader {
                src_port: conn.local_port,
                dest_port: conn.remote_port,
                seq_number: conn.seq_num,
                ack_number: conn.ack_num,
                data_offset: 0x50,
                flags: TCP_FLAG_FIN | TCP_FLAG_ACK,
                window_size: conn.window_size as u16,
                checksum: 0,
                urgent_ptr: 0,
            };
            use crate::ip::ip_send;
            let header_bytes = unsafe {
                core::slice::from_raw_parts(&tcp_header as *const _ as *const u8, 20)
            };
            let _ = ip_send(conn.remote_ip, crate::ip::IP_PROTOCOL_TCP, header_bytes);
            TCP_CONNECTIONS[conn_id] = None;
            Ok(())
        } else {
            Err(())
        }
    }
}

/// Send TCP data
pub fn tcp_send(conn_idx: usize, data: &[u8]) -> Result<(), ()> {
    unsafe {
        if conn_idx >= MAX_TCP_CONNECTIONS {
            return Err(());
        }
        
        if let Some(ref mut conn) = TCP_CONNECTIONS[conn_idx] {
            // Build TCP packet
            let mut tcp_header = TcpHeader {
                src_port: conn.local_port,
                dest_port: conn.remote_port,
                seq_number: conn.seq_num,
                ack_number: conn.ack_num,
                data_offset: 0x50,
                flags: TCP_FLAG_ACK | TCP_FLAG_PSH,
                window_size: conn.window_size as u16,
                checksum: 0,
                urgent_ptr: 0,
            };
            // Build packet
            let mut packet = [0u8; 1500];
            unsafe {
                core::ptr::copy_nonoverlapping(&tcp_header as *const _ as *const u8, packet.as_mut_ptr(), 20);
            }
            let data_len = data.len().min(1480);
            packet[20..20+data_len].copy_from_slice(&data[0..data_len]);
            // Send via IP layer
            use crate::ip::ip_send;
            let _ = ip_send(conn.remote_ip, crate::ip::IP_PROTOCOL_TCP, &packet[0..20+data_len]);
            Ok(())
        } else {
            Err(())
        }
    }
}

/// Receive TCP data
pub fn tcp_receive(conn_idx: usize, buffer: &mut [u8]) -> Result<usize, ()> {
    unsafe {
        if conn_idx >= MAX_TCP_CONNECTIONS {
            return Err(());
        }
        
        if let Some(_conn) = &TCP_CONNECTIONS[conn_idx] {
            // Receive from IP layer
            use crate::ip::ip_receive;
            let mut ip_buffer = [0u8; 1500];
            match ip_receive(&mut ip_buffer) {
                Ok((len, _src_ip, protocol)) => {
                    if protocol == crate::ip::IP_PROTOCOL_TCP && len >= 20 {
                        // Parse TCP header
                        let tcp_header = unsafe {
                            &*(ip_buffer.as_ptr() as *const TcpHeader)
                        };
                        // Copy data to buffer
                        let data_len = (len - 20).min(buffer.len());
                        buffer[0..data_len].copy_from_slice(&ip_buffer[20..20+data_len]);
                        Ok(data_len)
                    } else {
                        Ok(0)
                    }
                }
                Err(_) => Ok(0)
            }
        } else {
            Err(())
        }
    }
}

/// Handle TCP packet
pub fn tcp_handle_packet(buffer: &[u8], src_ip: u32) -> Result<(), ()> {
    // Parse TCP header
    if buffer.len() < 20 {
        return Err(());
    }
    let tcp_header = unsafe {
        &*(buffer.as_ptr() as *const TcpHeader)
    };
    
    // Find or create connection
    let mut conn_idx = None;
    unsafe {
        for i in 0..MAX_TCP_CONNECTIONS {
            if let Some(ref conn) = TCP_CONNECTIONS[i] {
                if conn.local_port == tcp_header.dest_port && 
                   conn.remote_ip == src_ip &&
                   conn.remote_port == tcp_header.src_port {
                    conn_idx = Some(i);
                    break;
                }
            }
        }
    }
    
    // Update connection state
    if let Some(idx) = conn_idx {
        unsafe {
            if let Some(ref mut conn) = TCP_CONNECTIONS[idx] {
                // Handle TCP state machine
                match conn.state {
                    TcpState::SynSent => {
                        if (tcp_header.flags & TCP_FLAG_SYN) != 0 && (tcp_header.flags & TCP_FLAG_ACK) != 0 {
                            conn.state = TcpState::Established;
                            conn.ack_num = tcp_header.seq_number + 1;
                        }
                    }
                    TcpState::Established => {
                        if (tcp_header.flags & TCP_FLAG_FIN) != 0 {
                            conn.state = TcpState::CloseWait;
                        } else if (tcp_header.flags & TCP_FLAG_ACK) != 0 {
                            conn.ack_num = tcp_header.seq_number;
                        }
                    }
                    _ => {}
                }
            }
        }
    }
    
    Ok(())
}

File: kernel/net/tcp.c
/**
 * @file tcp.c
 * @brief TCP protocol implementation
 */

#include "../include/net/tcp.h"
#include "../include/net/ip.h"
#include "../include/net/network.h"
#include "../include/kprintf.h"
#include "../include/debug.h"
#include "../include/mm/heap.h"
#include "../include/string.h"
#include "../include/sync/spinlock.h"

// TCP flag macros
#define TCP_FLAG_FIN 0x01
#define TCP_FLAG_SYN 0x02
#define TCP_FLAG_RST 0x04
#define TCP_FLAG_PSH 0x08
#define TCP_FLAG_ACK 0x10
#define TCP_FLAG_URG 0x20

// TCP connections
#define MAX_TCP_CONNECTIONS 32

static struct {
    tcp_connection_t* connections[MAX_TCP_CONNECTIONS];
    uint32_t next_seq_num;
    spinlock_t lock;
    bool initialized;
} tcp_state = {0};

// Forward declarations
extern error_code_t ip_send(uint32_t dest_ip, uint8_t protocol, void* data, size_t len);
extern net_device_t* network_find_device(const char* name);

/**
 * Calculate TCP checksum (pseudo-header + header + data)
 */
static uint16_t tcp_checksum(tcp_header_t* header, size_t tcp_len, uint32_t src_ip, uint32_t dest_ip) {
    // Pseudo-header for checksum
    struct {
        uint32_t src_ip;
        uint32_t dest_ip;
        uint8_t zero;
        uint8_t protocol;
        uint16_t tcp_length;
    } __attribute__((packed)) pseudo_header = {
        .src_ip = src_ip,
        .dest_ip = dest_ip,
        .zero = 0,
        .protocol = IP_PROTOCOL_TCP,
        .tcp_length = __builtin_bswap16(tcp_len)
    };
    
    uint32_t sum = 0;
    
    // Add pseudo-header
    uint16_t* words = (uint16_t*)&pseudo_header;
    for (int i = 0; i < 6; i++) {
        sum += __builtin_bswap16(words[i]);
    }
    
    // Add TCP header and data
    words = (uint16_t*)header;
    size_t word_count = tcp_len / 2;
    for (size_t i = 0; i < word_count; i++) {
        sum += __builtin_bswap16(words[i]);
    }
    
    // Add padding byte if odd length
    if (tcp_len % 2) {
        sum += ((uint8_t*)header)[tcp_len - 1] << 8;
    }
    
    while (sum >> 16) {
        sum = (sum & 0xFFFF) + (sum >> 16);
    }
    
    return __builtin_bswap16(~sum);
}

/**
 * Initialize TCP
 */
error_code_t tcp_init(void) {
    if (tcp_state.initialized) {
        return ERR_OK;
    }
    
    kinfo("Initializing TCP...\n");
    
    memset(tcp_state.connections, 0, sizeof(tcp_state.connections));
    tcp_state.next_seq_num = 1;  // Start at 1
    spinlock_init(&tcp_state.lock);
    tcp_state.initialized = true;
    
    kinfo("TCP initialized\n");
    return ERR_OK;
}

/**
 * Find TCP connection
 */
static tcp_connection_t* tcp_find_connection(uint32_t local_ip, uint16_t local_port, 
                                             uint32_t remote_ip, uint16_t remote_port) {
    spinlock_lock(&tcp_state.lock);
    
    for (int i = 0; i < MAX_TCP_CONNECTIONS; i++) {
        tcp_connection_t* conn = tcp_state.connections[i];
        if (conn &&
            conn->local_ip == local_ip &&
            conn->local_port == local_port &&
            conn->remote_ip == remote_ip &&
            conn->remote_port == remote_port) {
            spinlock_unlock(&tcp_state.lock);
            return conn;
        }
    }
    
    spinlock_unlock(&tcp_state.lock);
    return NULL;
}

/**
 * Create TCP connection
 */
tcp_connection_t* tcp_create_connection(uint32_t local_ip, uint16_t local_port, 
                                       uint32_t remote_ip, uint16_t remote_port) {
    if (!tcp_state.initialized) {
        tcp_init();
    }
    
    // Check if connection already exists
    tcp_connection_t* existing = tcp_find_connection(local_ip, local_port, remote_ip, remote_port);
    if (existing) {
        return existing;
    }
    
    // Allocate connection
    tcp_connection_t* conn = (tcp_connection_t*)kmalloc(sizeof(tcp_connection_t));
    if (!conn) {
        return NULL;
    }
    
    memset(conn, 0, sizeof(tcp_connection_t));
    conn->local_ip = local_ip;
    conn->local_port = local_port;
    conn->remote_ip = remote_ip;
    conn->remote_port = remote_port;
    conn->state = TCP_STATE_CLOSED;
    conn->seq_num = tcp_state.next_seq_num++;
    conn->ack_num = 0;
    conn->window_size = 4096;
    conn->receive_buffer_size = 4096;
    conn->receive_buffer = kmalloc(conn->receive_buffer_size);
    conn->receive_buffer_pos = 0;
    
    // Add to connection list
    spinlock_lock(&tcp_state.lock);
    for (int i = 0; i < MAX_TCP_CONNECTIONS; i++) {
        if (!tcp_state.connections[i]) {
            tcp_state.connections[i] = conn;
            break;
        }
    }
    spinlock_unlock(&tcp_state.lock);
    
    return conn;
}

/**
 * Send TCP packet
 */
error_code_t tcp_send_packet(tcp_connection_t* conn, uint8_t flags, void* data, size_t data_len) {
    if (!conn) {
        return ERR_INVALID_ARG;
    }
    
    size_t header_size = sizeof(tcp_header_t);
    size_t packet_size = header_size + data_len;
    
    tcp_header_t* tcp = (tcp_header_t*)kmalloc(packet_size);
    if (!tcp) {
        return ERR_OUT_OF_MEMORY;
    }
    
    memset(tcp, 0, packet_size);
    
    // Build TCP header
    tcp->src_port = conn->local_port;
    tcp->dest_port = conn->remote_port;
    tcp->seq_number = __builtin_bswap32(conn->seq_num);
    tcp->ack_number = __builtin_bswap32(conn->ack_num);
    tcp->data_offset = (header_size / 4) << 4;  // Data offset in 32-bit words
    tcp->flags = flags;
    tcp->window_size = __builtin_bswap16(conn->window_size);
    tcp->checksum = 0;
    tcp->urgent_ptr = 0;
    
    // Copy data
    if (data && data_len > 0) {
        memcpy(tcp->options + 0, data, data_len);
    }
    
    // Calculate checksum
    tcp->checksum = tcp_checksum(tcp, packet_size, conn->local_ip, conn->remote_ip);
    
    // Send via IP
    error_code_t err = ip_send(conn->remote_ip, IP_PROTOCOL_TCP, tcp, packet_size);
    
    // Update sequence number
    if (data_len > 0 || (flags & TCP_FLAG_SYN)) {
        conn->seq_num += data_len + ((flags & TCP_FLAG_SYN) ? 1 : 0);
    }
    
    kfree(tcp);
    return err;
}

/**
 * Send data on TCP connection
 */
error_code_t tcp_send(tcp_connection_t* conn, void* data, size_t len) {
    if (!conn || !data || len == 0) {
        return ERR_INVALID_ARG;
    }
    
    if (conn->state != TCP_STATE_ESTABLISHED) {
        return ERR_INVALID_STATE;
    }
    
    return tcp_send_packet(conn, TCP_FLAG_ACK | TCP_FLAG_PSH, data, len);
}

/**
 * Receive data from TCP connection
 */
error_code_t tcp_receive(tcp_connection_t* conn, void* buffer, size_t* len) {
    if (!conn || !buffer || !len) {
        return ERR_INVALID_ARG;
    }
    
    if (conn->state != TCP_STATE_ESTABLISHED) {
        return ERR_INVALID_STATE;
    }
    
    if (conn->receive_buffer_pos == 0) {
        return ERR_NOT_FOUND;  // No data available
    }
    
    size_t copy_len = (*len < conn->receive_buffer_pos) ? *len : conn->receive_buffer_pos;
    memcpy(buffer, conn->receive_buffer, copy_len);
    
    // Shift remaining data
    if (conn->receive_buffer_pos > copy_len) {
        memmove(conn->receive_buffer, 
                (uint8_t*)conn->receive_buffer + copy_len,
                conn->receive_buffer_pos - copy_len);
    }
    
    conn->receive_buffer_pos -= copy_len;
    *len = copy_len;
    
    return ERR_OK;
}

/**
 * Handle incoming TCP packet
 */
error_code_t tcp_handle_packet(void* buffer, size_t len, uint32_t src_ip) {
    if (!tcp_state.initialized || !buffer || len < sizeof(tcp_header_t)) {
        return ERR_INVALID_ARG;
    }
    
    tcp_header_t* tcp = (tcp_header_t*)buffer;
    
    uint16_t src_port = __builtin_bswap16(tcp->src_port);
    uint16_t dest_port = __builtin_bswap16(tcp->dest_port);
    uint32_t seq_num = __builtin_bswap32(tcp->seq_number);
    uint32_t ack_num = __builtin_bswap32(tcp->ack_number);
    uint8_t flags = tcp->flags;
    
    // Find connection
    net_device_t* device = network_find_device("eth0");
    if (!device || !device->up) {
        return ERR_DEVICE_NOT_FOUND;
    }
    
    tcp_connection_t* conn = tcp_find_connection(device->ip_address, dest_port, src_ip, src_port);
    
    if (!conn && (flags & TCP_FLAG_SYN)) {
        // New connection - SYN received
        // Check for listener
        tcp_connection_t* listener = tcp_find_connection(0, dest_port, 0, 0); // Listener has 0 remote IP/Port
        // Actually listeners might have local_ip set or 0 (INADDR_ANY).
        // Let's search for listener with matching local port and state LISTEN
        if (!listener) {
            spinlock_lock(&tcp_state.lock);
            for (int i = 0; i < MAX_TCP_CONNECTIONS; i++) {
                tcp_connection_t* c = tcp_state.connections[i];
                if (c && c->state == TCP_STATE_LISTEN && c->local_port == dest_port) {
                    listener = c;
                    break;
                }
            }
            spinlock_unlock(&tcp_state.lock);
        }
        
        if (listener) {
            // Create connection in SYN_RECEIVED state
            conn = tcp_create_connection(device->ip_address, dest_port, src_ip, src_port);
            if (conn) {
                conn->state = TCP_STATE_SYN_RECEIVED;
                conn->ack_num = seq_num + 1;
                
                // Send SYN-ACK
                tcp_send_packet(conn, TCP_FLAG_SYN | TCP_FLAG_ACK, NULL, 0);
                
                // For simplified TCP, assume established after SYN-ACK sent (skip 3-way handshake completion wait)
                // In real TCP, we wait for ACK.
                // But for now, let's just mark it established and queue it.
                conn->state = TCP_STATE_ESTABLISHED;
                
                // Add to listener's pending queue
                spinlock_lock(&tcp_state.lock);
                if (listener->pending_tail) {
                    listener->pending_tail->next_pending = (struct tcp_connection*)conn;
                    listener->pending_tail = (struct tcp_connection*)conn;
                } else {
                    listener->pending_head = (struct tcp_connection*)conn;
                    listener->pending_tail = (struct tcp_connection*)conn;
                }
                conn->next_pending = NULL;
                spinlock_unlock(&tcp_state.lock);
            }
            return ERR_OK;
        }
        
        // No listener, reject
        // Send RST?
        return ERR_NOT_FOUND;
    }
    
    if (!conn) {
        return ERR_NOT_FOUND;
    }
    
    // Update acknowledgment number
    if (flags & TCP_FLAG_ACK) {
        if (ack_num > conn->seq_num) {
            // Data acknowledged
        }
    }
    
    // Handle different flags
    if (flags & TCP_FLAG_FIN) {
        // Connection close
        if (conn->state == TCP_STATE_ESTABLISHED) {
            conn->state = TCP_STATE_CLOSE_WAIT;
            conn->ack_num = seq_num + 1;
            // Send ACK
            tcp_send_packet(conn, TCP_FLAG_ACK, NULL, 0);
            // Send FIN-ACK
            tcp_send_packet(conn, TCP_FLAG_FIN | TCP_FLAG_ACK, NULL, 0);
            conn->state = TCP_STATE_LAST_ACK;
        }
    }
    
    if (flags & TCP_FLAG_ACK && conn->state == TCP_STATE_LAST_ACK) {
        // FIN acknowledged
        tcp_close(conn);
        return ERR_OK;
    }
    
    // Extract data
    size_t data_offset = ((tcp->data_offset >> 4) & 0x0F) * 4;
    if (data_offset < sizeof(tcp_header_t)) {
        data_offset = sizeof(tcp_header_t);
    }
    
    if (len > data_offset) {
        size_t data_len = len - data_offset;
        void* data = (uint8_t*)buffer + data_offset;
        
        // Store in receive buffer
        if (conn->receive_buffer_pos + data_len <= conn->receive_buffer_size) {
            memcpy((uint8_t*)conn->receive_buffer + conn->receive_buffer_pos, data, data_len);
            conn->receive_buffer_pos += data_len;
            conn->ack_num = seq_num + data_len;
            
            // Send ACK
            tcp_send_packet(conn, TCP_FLAG_ACK, NULL, 0);
        }
    }
    
    return ERR_OK;
}

/**
 * Listen on a port
 */
error_code_t tcp_listen(uint16_t port) {
    if (!tcp_state.initialized) {
        tcp_init();
    }
    
    // Check if already listening
    spinlock_lock(&tcp_state.lock);
    for (int i = 0; i < MAX_TCP_CONNECTIONS; i++) {
        tcp_connection_t* c = tcp_state.connections[i];
        if (c && c->state == TCP_STATE_LISTEN && c->local_port == port) {
            spinlock_unlock(&tcp_state.lock);
            return ERR_ALREADY_EXISTS;
        }
    }
    spinlock_unlock(&tcp_state.lock);
    
    // Create listener connection
    // Use 0.0.0.0 (INADDR_ANY) for local IP to listen on all interfaces
    // But tcp_create_connection takes specific IP.
    // We can use 0 to indicate ANY?
    // Let's use 0.
    tcp_connection_t* conn = tcp_create_connection(0, port, 0, 0);
    if (!conn) {
        return ERR_OUT_OF_MEMORY;
    }
    
    conn->state = TCP_STATE_LISTEN;
    return ERR_OK;
}

/**
 * Accept a connection
 */
tcp_connection_t* tcp_accept(uint16_t port) {
    // Find listener
    tcp_connection_t* listener = NULL;
    spinlock_lock(&tcp_state.lock);
    for (int i = 0; i < MAX_TCP_CONNECTIONS; i++) {
        tcp_connection_t* c = tcp_state.connections[i];
        if (c && c->state == TCP_STATE_LISTEN && c->local_port == port) {
            listener = c;
            break;
        }
    }
    
    if (!listener) {
        spinlock_unlock(&tcp_state.lock);
        return NULL;
    }
    
    // Check pending queue
    if (listener->pending_head) {
        tcp_connection_t* conn = (tcp_connection_t*)listener->pending_head;
        listener->pending_head = conn->next_pending;
        if (!listener->pending_head) {
            listener->pending_tail = NULL;
        }
        conn->next_pending = NULL;
        spinlock_unlock(&tcp_state.lock);
        return conn;
    }
    
    spinlock_unlock(&tcp_state.lock);
    return NULL; // No pending connections (would block in blocking mode)
}

/**
 * Close TCP connection
 */
error_code_t tcp_close(tcp_connection_t* conn) {
    if (!conn) {
        return ERR_INVALID_ARG;
    }
    
    if (conn->state == TCP_STATE_ESTABLISHED) {
        // Send FIN
        tcp_send_packet(conn, TCP_FLAG_FIN | TCP_FLAG_ACK, NULL, 0);
        conn->state = TCP_STATE_FIN_WAIT_1;
    }
    
    // Remove from connection list
    spinlock_lock(&tcp_state.lock);
    for (int i = 0; i < MAX_TCP_CONNECTIONS; i++) {
        if (tcp_state.connections[i] == conn) {
            tcp_state.connections[i] = NULL;
            break;
        }
    }
    spinlock_unlock(&tcp_state.lock);
    
    if (conn->receive_buffer) {
        kfree(conn->receive_buffer);
    }
    kfree(conn);
    
    return ERR_OK;
}

